### DISCUSSION TITLE
Unable to log in to self-hosted Outline install due to "State not return in OAuth flow" error

### DISCUSSION THREAD

**Discussion by SapphoSys** [outline/outline#7316] [Self Hosting] [answered]
Hi! Yesterday, I was unable to log in to my Outline instance with Discord due to an authentication error. A friend of mine informed me of the same issues a few hours ago. Which was really odd, since this exact setup was working just fine days, weeks, and months prior.

**To Reproduce**
1. Configure a new Outline install
2. Use Discord as an OIDC provider

**Expected behavior**
You log in to your self-hosted instance of Outline with your Discord credentials.

**Actual behavior**
Outline reports a "State not return in OAuth flow" error. (BadRequestError: State not return in OAuth flow at OAuthStateMismatchError)

**Error log**
https://haste.erisa.uk/dufilaqere.swift

**Additional Information**
The Discord application's redirect URL is set to https://mycustomdomain.com/auth/oidc.callback.

Here's Outline's `compose.yml` environment variables:
https://haste.erisa.uk/zuwumaruji.bash

Here's Outline's `.env` file:
https://haste.erisa.uk/iduqicusuv.ini

**Screenshots**
![image](https://github.com/user-attachments/assets/58197520-56ba-4485-94f5-a6769c168e8d)

**Outline (please complete the following information):**
- Install: Self-hosted
- Version: 
```
$ docker inspect --format='{{index .RepoDigests 0}}' 33086a16cea1
outlinewiki/outline@sha256:d317b2e03b8e61ea013f27c08f83e23a9986cec44545fd2370e4f7b6c0213e11
```

**Desktop (please complete the following information):**
 - OS: Microsoft Windows 11 Version 23H2 (OS Build 22631.3880)
 - Browser: Microsoft Edge
 - Version: 126.0.2592.113 (Official build) (64-bit)

**Mobile (please complete the following information):**
 - Device: OnePlus 9 5G
 - OS: Android 14
 - Browser: Microsoft Edge

**Comment by tommoor** [comment]
Take a look at some of the previous discussions around this:

https://github.com/outline/outline/discussions/2089#discussioncomment-667685

**Reply by SapphoSys** [reply]
Unfortunately, I'm using Caddy instead of nginx, so the answer provided in the discussion is not applicable for me.

Here's how my Outline instance is currently set up in Caddyfile:
```
mycustomdomain.com {
  reverse_proxy http://outline:3000
}
```

**Reply by SapphoSys** [reply]
If there's any other way you're able to help, I'd appreciate it a lot. Thank you so much in advance! ^^

**Reply by tommoor** [reply]
There is a cookie with the name `state` set on the server when starting the OAuth flow, the value of the cookie is checked when you return back from the authentication provider – if this error appears it's because the cookie was not found. You will have to inspect the requests and responses that happen during the OAuth flow to see the cookie being set, most likely the incorrect domain is used somewhere so it is not accessible on return.

**Reply by tpietsch-rcp** [reply]
I might be seeing a similar issue here - using keycloak for self hosting and keycloak generates this redirect back

```bash
https://outline.kaledo.ai/auth/oidc.callback?state=794e2331bc0de207&session_state=dc8c6683-d6df-4dc4-9edb-682ca5e9a1f9&iss=https%3A%2F%2Fcloak.kaledo.ai%2Frealms%2Foutline&code=0d6674c1-4807-4d57-97d1-c413101727ac.dc8c6683-d6df-4dc4-9edb-682ca5e9a1f9.fcdd16f3-c9ac-4c59-b983-0881c0e73575
```
Which includes the state param `state=794e2331bc0de207`

But I dont see any cookie being set by keycloak/outline and get the same error about state not being set
```javascript
{"error":"State not return in OAuth flow","level":"error","message":"Error during authentication","stack":"BadRequestError: State not return in OAuth flow\n    at OAuthStateMismatchError (/opt/outline/build/server/errors.js:147:34)\n    at StateStore.verify (/opt/outline/build/server/utils/passport.js:46:61)\n    at OAuth2Strategy.authenticate (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:222:26)\n    at _passportOauth.Strategy.authenticate (/opt/outline/build/plugins/oidc/server/auth/oidc.js:32:26)\n    at attempt (/opt/outline/node_modules/@outlinewiki/koa-passport/node_modules/passport/lib/middleware/authenticate.js:369:16)\n    at authenticate (/opt/outline/node_modules/@outlinewiki/koa-passport/node_modules/passport/lib/middleware/authenticate.js:370:7)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:194:7\n    at new Promise (<anonymous>)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:193:12\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:143:7\n    at new Promise (<anonymous>)\n    at passportAuthenticate (/opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:107:15)\n    at passportMiddleware (/opt/outline/build/server/middlewares/passport.js:75:7)\n    at /opt/outline/node_modules/dd-trace/packages/datadog-instrumentations/src/koa.js:88:57\n    at passportMiddleware (/opt/outline/node_modules/dd-trace/packages/datadog-shimmer/src/shimmer.js:26:21)\n    at dispatch (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/koa-router/lib/router.js:346:16\n    at dispatch (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)\n    at next (/opt/outline/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)\n    at /opt/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:60:14\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"}
```

**Reply by tommoor** [reply]
> Which includes the state param

This error is referring to the cookie, not the `state` query string parameter so it's a bit confusing. The cookie is set as part of forwarding to your OIDC provider so you should inspect that request to see what's happening

**Comment by bdr99** [comment, marked as answer]
If anyone is having this issue, make sure your system clock is set correctly. My clock was set a few hours in the future, which was causing my browser to expire the `state` cookie immediately after it was set.

**Comment by blitzdose** [comment]
Duplicate https://github.com/outline/outline/discussions/8322#discussioncomment-13994440
